### PR TITLE
JetBrains: Improve search box UI/UX

### LINKED ISSUE #36702: JetBrains: Search box styling
Including the Monaco dropdown lists

### LINKED ISSUE #37199: JetBrains: Typing inside the search input box is really slow
- We currently re-render the search result list on every keystroke
- We also store a saved search on every keystroke

Will be fixed in #37198

### PR DESCRIPTION
**Pull request by philipp-spiess**
Fixes #36702 
Fixes #37199

__⚠️ I recommend the code reviewer to review commit-by-commit.__

This PR applies a bunch of changes to the JetBrains search box trying to mimic the [Figma](https://www.figma.com/file/nmNPZGYCoGIUN0YMylmPTT/JetBrains-plugin-MLP?node-id=246%3A11) as good as possible.

Customization this part of the design is complicated since we do make heavy use of existing web components so for any components that need heaver customization, we have to fork the component into the JetBrains repo which dramatically increases maintainance cost as we now have two of the same components to maintain. To keep this as low as possible I tried to make some adjustments to get away with only forking two components (`<SearchBox>` and `<Toggles>`).

Here is what is currently missing from the Figma that I am aware of:

- I decided to not include an icon in the context menu dropdown since this would require me to fork another module. If we can avoid it, I'd strongly suggest we do.
- The individual drop downs (context menu and search specific ones) aren't styled at all yet. They are following two separate implementations that will be hard to match. Instead, for now, I suggest we use the default wildcard styling.

**Known issues (they already happen on `main`):**

- https://github.com/sourcegraph/sourcegraph/issues/37210

**Other issues this fixes:**

- I found out why the cursor has a weird blue border.
- I noticed that typing is _really_ slow. Turns out we updated every result in the result list on every keystroke 😐 We also saved the previous search query after every keystroke. With these two issues fixed we now have a vastly better typing experience.
- The focus ring now looks much better:

Screenshot 2022-06-14 at 16 20 01

## Test plan

This was tested manually. There isn't a lot of interactions that we need to take care here so I decided to keep the test plan simple and include two screenshots from dark and light mode instead:

### With focus

Screenshot 2022-06-14 at 16 51 47
Screenshot 2022-06-14 at 16 51 39

### Without focus

Screenshot 2022-06-14 at 16 56 12
Screenshot 2022-06-14 at 16 55 54

## App preview:

- [Web](https://sg-web-ps-jetbrains-searchbox.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-lyvqxeveig.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by philipp-spiess** [comment]
@jjinnii Hmmmmmmmmm I’m not sure 😛  [Of course!](https://drive.google.com/file/d/1iIorQyfNhKdkjsAlkOW_KSqJrcW3gr0X/view?usp=sharing)

**Review by vdavid — APPROVED** [review]
I went through the whole code, then commit by commit, then scrolled through the whole code again, and it all seems to be nice to me.

Structure:
I think it was a good call to try to minimize the number of forked components. My alternative idea would be to modify the original component, make it customizable, and use it both on the website and here. I think it'd be more maintainable because less duplication, but
1. I guess it was not viable because otherwise you'd have done this
2. Even if it's viable but this was quicker to do, I'm okay with this debt.

From a UI point of view:
It looks very nice!
The only thing that stands out is the blue focus glow around the text box which doesn't exist anywhere in the IntelliJ UI, so that seems to be the strongest giveaway for me that this is not a native component but a webview. I'd consider removing that.

File: client/jetbrains/webview/src/index.scss [hunk 1]
```diff
@@ -1 +1,15 @@
 @import '../../../branded/src/global-styles/index.scss';
+
+.monaco-mouse-cursor-text {
+    /* Hack: remove focus ring around Monaco editor cursor */
```
[the remark this answers is not in the document]
Comment (philipp-spiess) [reply client/jetbrains/webview/src/index.scss:4]: Haha it took me long to figure out what's going on but once I found it I also found this Lol: https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+Hack:+remove+focus+ring+around+Monaco+editor+cursor+&patternType=literal

[the remark this answers is not in the document]
Comment (vdavid) [reply client/jetbrains/webview/src/index.scss:4]: Now we have four of them :D

**Comment by philipp-spiess** [comment]
> I think it was a good call to try to minimize the number of forked components. My alternative idea would be to modify the original component, make it customizable, and use it both on the website and here. I think it'd be more maintainable because less duplication, but
> 
> I guess it was not viable because otherwise you'd have done this
> Even if it's viable but this was quicker to do, I'm okay with this debt.

Thanks this is great feedback! Yeah when I started with the `SearchBox` component my reasoning was that the component itself is fairly small and only really defines UI/UX anyway and since I needed to change quite a lot of things, it appeared much easier to fork it. I later expanded the pattern to the `Toggles` component but I now think that this might have been a better place to instead add this into the existing component. The changes there was mostly a reordering of the separator and the addition of an additional button. If you feel strongly about this I can try to upstream those changes so that we only need to have our own search box (which makes sense since we have a vastly different search box UI as well). WDYT?

> The only thing that stands out is the blue focus glow around the text box which doesn't exist anywhere in the IntelliJ UI, so that seems to be the strongest giveaway for me that this is not a native component but a webview. I'd consider removing that.

Yeah absolutely. I didn't touch it for the search box since it was also used for all other buttons in our UI yet but I also noticed that it looks off. 

Native JetBrains does not show an outline when the search input is focused but does show one when you tab to a button:

Screenshot 2022-06-15 at 11 59 08
Screenshot 2022-06-15 at 11 59 07

**Comment by jjinnii** [comment]
@philipp-spiess 

- On focus, I think following the design principle we've been keeping to, I agree with David that I think it'd be more consistent to remove the focus area! The blue outline should be maintained though if a user is using keyboard navigation to utilize the buttons like you called out. 
- The typing doesn't feel incredibly slow to me, but I'm also not a developer 😄 
- Sometimes when I open the plugin, the search bar seems to be cut off. I think this is less of a problem when we remove the focus outline, but the icons are being cut off as well.
Screen Shot 2022-06-15 at 6 48 45 PM


I think it feels really great, and it's an amazing improvement!

**Review by jjinnii — APPROVED** [review]

**Comment by philipp-spiess** [comment]
> * On focus, I think following the design principle we've been keeping to, I agree with David that I think it'd be more consistent to remove the focus area! The blue outline should be maintained though if a user is using keyboard navigation to utilize the buttons like you called out.

I created a follow-up issue for that: #37285

> The typing doesn't feel incredibly slow to me, but I'm also not a developer 😄

This is because I already fixed the issue in this PR 🙈   Sorry for not making it clear in the description but the heading reads "Other issues **this fixes:**"

> Sometimes when I open the plugin, the search bar seems to be cut off. I think this is less of a problem when we remove the focus outline, but the icons are being cut off as well.

Does this happen when you just load it or only after you manually resize it? I noticed a similar issue that only happens when we shrink the window and it seems like Monaco (the text editor that we use) doesn't properly update for some reason :( I created a ticket for it as well: https://github.com/sourcegraph/sourcegraph/issues/37210

If this is another issue I'd probably make another issue even if those have the same root.

> I think it feels really great, and it's an amazing improvement!

❤️

**Comment by philipp-spiess** [comment]
This PR is ready to merge based on design and engineering. I’m still waiting on a [response from legal](https://sourcegraph.slack.com/archives/C01GJFE26BX/p1655293315005859) but if this looks good, feel free to merge this on my behalf @vdavid 🙂

**Comment by vdavid** [comment]
Got response from legal, we're good!

**Pull request merged**